propagate.go

```diff
@@ -0,0 +1,118 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel"
+
+	"github.com/cyverse-de/go-mod/restutils"
+	"github.com/cyverse-de/group-propagator/client/datainfo"
+	"github.com/cyverse-de/group-propagator/client/groups"
+)
+
+// To propagate a group:
+// * Fetch group details and members via iplant-groups
+//   -> get a model.GrouperGroup and model.GrouperGroupMembers, probably
+// * Determine iRODS group name (@grouper-<GrouperGroup.ID>)
+// * Create or update group with proper membership list via data-info, potentially validating users/etc.
+
+type Propagator struct {
+	groupsClient *groups.GroupsClient
+	groupPrefix  string
+
+	dataInfoClient *datainfo.DataInfoClient
+}
+
+func NewPropagator(groupsClient *groups.GroupsClient, groupPrefix string, dataInfoClient *datainfo.DataInfoClient) *Propagator {
+	if groupPrefix == "" {
+		groupPrefix = "@grouper-"
+	}
+
+	return &Propagator{
+		groupsClient:   groupsClient,
+		groupPrefix:    groupPrefix,
+		dataInfoClient: dataInfoClient,
+	}
+}
+
+func (p *Propagator) getGroupMembers(ctx context.Context, groupName string) ([]string, error) {
+	ctx, span := otel.Tracer(otelName).Start(ctx, "getGroupMembers")
+	defer span.End()
+
+	var m []string
+
+	members, err := p.groupsClient.GetGroupMembers(ctx, groupName)
+	if err != nil {
+		return m, errors.Wrapf(err, "Failed fetching Grouper group members for %s", groupName)
+	}
+
+	for _, member := range members.Members {
+		if member.SourceID == "ldap" {
+			m = append(m, member.ID)
+		} else if member.SourceID == "g:gsa" {
+			// this is a group that is a member of a group
+			submem, err := p.getGroupMembers(ctx, member.Name)
+			if err != nil {
+				return m, errors.Wrapf(err, "Failed recursing to fetch members of %s", member.Name)
+			}
+			m = append(m, submem...)
+		} else {
+			log.Errorf("Could not add group member %+v", member)
+		}
+	}
+
+	return m, nil
+}
+
+func (p *Propagator) PropagateGroupById(ctx context.Context, groupID string) error {
```

Comment:
This function, and the things it calls, are largely the real core functionality of the service.